mvpa2/algorithms/statis.py

```diff
@@ -0,0 +1,142 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the PyMVPA package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Transformation of individual feature spaces into a compromise common space
+
+The :class:`Statis` class in this module implements an algorithm
+published in Abdi, H., Williams, L.J., Valentin, D., & Bennani-Dosse, M. (2012). STATIS
+and DISTATIS: Optimum multi-table principal component analysis and three way
+metric multidimensional scaling. Wiley Interdisciplinary Reviews:
+Computational Statistics, 4, 124-167.
+
+"""
+
+import numpy as np
+from mvpa2.base.types import is_datasetlike
+from mvpa2.base.state import ClassWithCollections
+from mvpa2.mappers.zscore import ZScoreMapper
+from mvpa2.mappers.staticprojection import StaticProjectionMapper
+
+def normalize_ds(ds):
+    """
+    Normalizes input dataset by
+    1) zscoring each column
+    2) making sum of squares to one (Frobenius norm = 1)
+    :param ds:
+    :return ds:
+    """
+    if ds.samples.dtype == int:
+        ds.samples = ds.samples.astype('float')
+    zmapper = ZScoreMapper(chunks_attr=None)
+    zmapper.train(ds)
+    ds = zmapper.forward(ds)
+    ds.samples = ds.samples/np.sqrt(((ds.samples**2).sum()))
+    return ds
+
+
+def get_Rv(dss):
+    """
+    Computes Rv matrix: Cosine correlation of RSMs across datasets
+    :param dss: List of datasets, typically one per subject, with
+     matching samples
+    :return: Rv
+    """
+    z = np.vstack([np.dot(sd.samples, sd.samples.T).flatten() for sd in dss])
+    return z, np.dot(z, z.T)
+
+
+def get_eig(h):
+    """
+    Helper function to return sorted and sign flipped
+    Eigen decomposition.
+    :param C: Symmetric matrix
+    :return: eigen values, eigen vectors
+    """
+    evals, evecs = np.linalg.eigh(h)
+    # To deal with numerical errors
+    evals[evals<0] = 0
+    evecs = evecs[:, np.argsort(evals)[-1::-1]]
```

Comment:
how about storing the result from np.argsort, and use it to index evals and evecs? In that way there is only a single sort operation.